cluster/src/main/java/org/apache/iotdb/cluster/query/ClusterPlanExecutor.java

```diff
@@ -81,22 +99,172 @@ public QueryDataSet processQuery(PhysicalPlan queryPlan, QueryContext context)
   }
 
   @Override
-  protected List<String> getNodesList(String schemaPattern, int level) {
-    // TODO-Cluster: enable meta queries
-    throw new UnsupportedOperationException("Not implemented");
-    //return metaGroupMember.getNodeList(schemaPattern, level);
+  protected List<String> getNodesList(String schemaPattern, int level)
+      throws IOException, TException, InterruptedException, MetadataException {
+    ConcurrentSkipListSet<String> nodeSet = new ConcurrentSkipListSet<>(
+        MManager.getInstance().getNodesList(schemaPattern, level));
+
+    ExecutorService pool = new ScheduledThreadPoolExecutor(THREAD_POOL_SIZE);
+
+    for (PartitionGroup group : metaGroupMember.getPartitionTable().getLocalGroups()) {
+      Node header = group.getHeader();
+      if (header.equals(metaGroupMember.getThisNode())) {
+        continue;
+      }
+      pool.submit(() -> {
+        GetNodesListHandler handler = new GetNodesListHandler();
+        AtomicReference<List<String>> response = new AtomicReference<>(null);
+        handler.setResponse(response);
+
+        for (Node node : group) {
+          try {
+            DataClient client = metaGroupMember.getDataClient(node);
+            handler.setContact(node);
+            synchronized (response) {
+              if (client != null) {
+                client.getNodeList(header, schemaPattern, level, handler);
+                response.wait(connectionTimeoutInMS);
+              }
+            }
+            List<String> paths = response.get();
+            if (paths != null) {
+              nodeSet.addAll(paths);
+              break;
+            }
+          } catch (IOException e) {
+            logger.error("Failed to connect to node: {}", node, e);
+          } catch (TException e) {
+            logger.error("Error occurs when getting node lists in node {}.", node, e);
+          } catch (InterruptedException e) {
+            logger.error("Interrupted when getting node lists in node {}.", node, e);
+            Thread.currentThread().interrupt();
+          }
+        }
```

Comment:
It would be better to leave a log here if you fail to get result from all nodes in the group.

Reply:
Yes, it would be more clear.